### PR TITLE
feat(termination)!: make container termination timeout configurable

### PR DESCRIPTION
**Pull request by moogacs**
## What does this PR do?

- extend `Terminate()` method with optional configuration for timeout configuration. 
- moved the `terminationOptions` under container to give the ability to make timeout, etc  configurable via a functional option.
- moved cleaning up under `Terminate()` function.

## Why is it important?

- `Terminate()` became broken with latest release version with latest release `0.34.0` `Terminate()` require handle for errors `"is already in progress"`
- this is CleanupContainer method signature and to use it in down streams that will require  many refactors because it expects `testing.TB`  
```
func CleanupContainer(tb testing.TB, ctr Container, options ...TerminateOption) {
```

so  with this change it will give the ability to configure the timeout and move the cleanup logic under termination 

## Related issues

- Closes https://github.com/testcontainers/testcontainers-go/issues/2917

a unit test added and shall be tested with all the existed pipelines

### CODE REVIEW

File: container.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ type Container interface {
 	Stop(context.Context, *time.Duration) error                                    // stop the container
 
 	// Terminate stops and removes the container and its image if it was built and not flagged as kept.
-	Terminate(ctx context.Context) error
+	Terminate(ctx context.Context, opts ...TerminateOption) error
```
Comment (mdelapenya) [inline container.go:53]: As this is breaking change, the PR title must be updated to `feat(termination)!: ...`

See https://golang.testcontainers.org/contributing/#submitting-your-changes

Comment (moogacs) [reply container.go:53]: thank you sure, gonna update it. i didn't think of it as breaking change, given the variadic function will make `opts ` optional

Comment (mdelapenya) [reply container.go:53]: Indeed, it could slip into a breaking change in two cases: somebody implementing their own container, i.e. modules, as you already spotted in the etcd module, and if somebody has the Terminate function assigned to a variable (see https://go.dev/play/p/5LsWTXgv_iU). Very rare thought

Comment (moogacs) [reply container.go:53]: gotcha, good point 👍🏼

File: docker.go [hunk 2]
```diff
@@ -296,19 +297,33 @@ func (c *DockerContainer) Stop(ctx context.Context, timeout *time.Duration) erro
...
-func (c *DockerContainer) Terminate(ctx context.Context) error {
-	// ContainerRemove hardcodes stop timeout to 3 seconds which is too short
-	// to ensure that child containers are stopped so we manually call stop.
-	// TODO: make this configurable via a functional option.
-	timeout := 10 * time.Second
-	err := c.Stop(ctx, &timeout)
+//
+// Default: timeout is 10 seconds.
+func (c *DockerContainer) Terminate(ctx context.Context, opts ...TerminateOption) error {
+	if len(opts) == 0 {
```
Comment (stevenh) [inline docker.go:317]: bug: this is fragile if other options are added.

**Review by stevenh — CHANGES_REQUESTED** [review]
Thanks for PR, seems like a good idea, unfortunately this approach doesn't work with an interface design, see comments for details.

File: cleanup.go [hunk 3]
```diff
@@ -54,41 +53,12 @@ func TerminateContainer(container Container, options ...TerminateOption) error {
...
-		ctx: context.Background(),
-	}
-
-	for _, opt := range options {
-		opt(c)
-	}
-
-	// TODO: Add a timeout when terminate supports it.
-	err := container.Terminate(c.ctx)
+	err := container.Terminate(context.Background(), options...)
```
[the remark this answers is not in the document]
Comment (moogacs) [reply cleanup.go:56]: you mean like convert Terminate to accept `opts` only ?

File: cleanup.go [hunk 4]
```diff
@@ -16,21 +15,21 @@ type terminateOptions struct {
 }
 
 // TerminateOption is a type that represents an option for terminating a container.
-type TerminateOption func(*terminateOptions)
+type TerminateOption func(*DockerContainer)
```
[the remark this answers is not in the document]
Comment (moogacs) [reply cleanup.go:18]: yeah, i had similar feeling, but wasn't sure if shall i introduce a separate time, when i was opening this PR i tried to make as minimal breaking change but seems it has to be. open for suggestions 

i like the idea of `maintain the dedicated type but export it and the fields`

File: cleanup.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (stevenh) [reply cleanup.go:56]: Ignore me I was confusing clean up and terminate, the former not having a context param.

File: cleanup.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (stevenh) [reply cleanup.go:18]: I initially thought the dedicated type would work, but that wont ensure that all Container implementations support the options, which I'm not a fan of as a consumer would be left wondering why their options didn't work.

[the remark this answers is not in the document]
Comment (moogacs) [reply cleanup.go:18]: the other thought i had before i come to this was introducing it in the `ContainerRequest`  and add getter in the `container` interface

[the remark this answers is not in the document]
Comment (stevenh) [reply cleanup.go:18]: I thought about setters too, but I think that would make it much harder to discover for users.

I can't think of any perfect solution ATM, something mull over.

[the remark this answers is not in the document]
Comment (moogacs) [reply cleanup.go]: i have made a change and revert it to make sure it uses `terminateOptions`  instead of the dependency on DockerContainer type. 

also added more test to show case the usage and the changes of configuration 

defaults of volumes is nil, timeout is 10   and context is passed already on `Terminate()` , lmk wdyt ?

File: modules/etcd/etcd.go [hunk 5]
```diff
@@ -29,12 +29,12 @@ type EtcdContainer struct {
...
 // Terminate terminates the etcd container, its child nodes, and the network in which the cluster is running
 // to communicate between the nodes.
-func (c *EtcdContainer) Terminate(ctx context.Context) error {
+func (c *EtcdContainer) Terminate(ctx context.Context, opts ...testcontainers.TerminateOption) error {
 	var errs []error
 
 	// child nodes has no other children
 	for i, child := range c.childNodes {
-		if err := child.Terminate(ctx); err != nil {
+		if err := child.Terminate(ctx, opts...); err != nil {
```
Comment (stevenh) [inline modules/etcd/etcd.go:37]: bug: not passed to the parent container below.

File: docker.go [hunk 6]
```diff
@@ -296,19 +296,41 @@ func (c *DockerContainer) Stop(ctx context.Context, timeout *time.Duration) erro
...
-	// ContainerRemove hardcodes stop timeout to 3 seconds which is too short
-	// to ensure that child containers are stopped so we manually call stop.
-	// TODO: make this configurable via a functional option.
-	timeout := 10 * time.Second
-	err := c.Stop(ctx, &timeout)
+//
+// Default: timeout is 10 seconds.
+func (c *DockerContainer) Terminate(ctx context.Context, opts ...TerminateOption) error {
+	defaultTimeout := 10 * time.Second
+	defaultOptions := &terminateOptions{
```
Comment (stevenh) [inline docker.go:324]: bug: for this to work for implementations outside of testcontainers package this would need to be exported alongside the fields.

File: docker.go [hunk 7]
```diff
@@ -343,6 +365,25 @@ func (c *DockerContainer) Terminate(ctx context.Context) error {
 	c.sessionID = ""
 	c.isRunning = false
 
+	// Remove additional volumes if any.
```
Comment (stevenh) [inline docker.go:368]: suggestion: can you add a TODO to simplify this when when we perform the client refactor please.

File: docker_test.go [hunk 8]
```diff
@@ -279,6 +279,27 @@ func TestContainerStateAfterTermination(t *testing.T) {
 		require.Nil(t, state, "expected nil container inspect.")
 	})
 
+	t.Run("Nil State after termination with timeout", func(t *testing.T) {
```
Comment (stevenh) [inline docker_test.go:282]: suggestion: we're standardising on short names with no spaces so a search from the failed sub test name can be successful.
```suggestion
	t.Run("termination-timeout", func(t *testing.T) {
```

File: docker_test.go [hunk 9]
```diff
@@ -1087,6 +1108,56 @@ func TestContainerCreationWithVolumeAndFileWritingToIt(t *testing.T) {
...
+					ContainerFilePath: "/hello.sh",
+				},
+			},
+			Mounts:     Mounts(VolumeMount(volumeName, "/data")),
+			Cmd:        []string{"bash", "/hello.sh"},
+			WaitingFor: wait.ForLog("done"),
+		},
+		Started: true,
+	})
+	require.NoError(t, err)
```
Comment (stevenh) [inline docker_test.go:1137]: bug: missing `ContainerCleanup` call before `NoError`.

File: docker_test.go [hunk 10]
```diff
@@ -1087,6 +1108,56 @@ func TestContainerCreationWithVolumeAndFileWritingToIt(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestContainerCreationWithVolumeCleaning(t *testing.T) {
+	absPath, err := filepath.Abs(filepath.Join(".", "testdata", "hello.sh"))
+	require.NoError(t, err)
```
Comment (stevenh) [inline docker_test.go:1113]: Just so you're aware (no need to change) we've been using embed for this type recently to its easier to reuse.
```go
//go:embed testdata/hello.sh
var helloSh []byte
```
Then use:
```go
ContainerFile{
    Reader: bytes.NewReader(helloSh),
    ContainerFilePath: "/hello.sh",
    FileMode: 0o755,
}
```

File: docker_test.go [hunk 11]
```diff
@@ -1087,6 +1108,56 @@ func TestContainerCreationWithVolumeAndFileWritingToIt(t *testing.T) {
...
+
+	// Create the container that writes into the mounted volume.
+	bashC, err := GenericContainer(ctx, GenericContainerRequest{
+		ProviderType: providerType,
+		ContainerRequest: ContainerRequest{
+			Image: "bash:5.2.26",
+			Files: []ContainerFile{
+				{
+					HostFilePath:      absPath,
+					ContainerFilePath: "/hello.sh",
```
Comment (stevenh) [inline docker_test.go:1128]: bug: Missing `FileMode` field which will cause unpredictable behaviour.

File: docker_test.go [hunk 12]
```diff
@@ -279,6 +279,27 @@ func TestContainerStateAfterTermination(t *testing.T) {
...
+		ctx := context.Background()
+		nginx, err := createContainerFn(ctx)
+		require.NoError(t, err)
+
+		err = nginx.Start(ctx)
+		require.NoError(t, err, "expected no error from container start.")
+
+		state, err := nginx.State(ctx)
+		require.NoError(t, err, "expected no error from container inspect.")
+		require.NotNil(t, state, "expected non nil container inspect.")
```
Comment (stevenh) [inline docker_test.go:292]: nit: I wouldn't include this as we're not testing the behaviour of the `State` call, similarly with the `Nil` check after `require.Error`.

**Review by stevenh — CHANGES_REQUESTED** [review]
Thanks for fleshing out what this looks like, I've done a review on how it stands and will discuss with @mdelapenya in the new year the pros and cons of functional options vs require args.

File: modules/etcd/etcd.go (hunk 5 above)
Comment (moogacs) [reply modules/etcd/etcd.go:37]: good catch, thanks

File: docker.go [hunk 13]
```diff
@@ -296,19 +296,40 @@ func (c *DockerContainer) Stop(ctx context.Context, timeout *time.Duration) erro
...
+	defaultOptions := &TerminateOptions{
+		Timeout: &DefaultTimeout,
+		Context: ctx,
+	}
+
+	for _, opt := range opts {
+		opt(defaultOptions)
+	}
+
+	err := c.Stop(defaultOptions.Context, defaultOptions.Timeout)
```
Comment (stevenh) [inline docker.go:332]: suggestion: use method as defined above
```suggestion
	err := c.Stop(options.Context(), options.StopTimeout())
```

File: docker.go [hunk 14]
```diff
@@ -296,19 +296,40 @@ func (c *DockerContainer) Stop(ctx context.Context, timeout *time.Duration) erro
...
+// Default: timeout is 10 seconds.
+func (c *DockerContainer) Terminate(ctx context.Context, opts ...TerminateOption) error {
+	defaultOptions := &TerminateOptions{
+		Timeout: &DefaultTimeout,
+		Context: ctx,
+	}
+
+	for _, opt := range opts {
+		opt(defaultOptions)
+	}
```
Comment (stevenh) [inline docker.go:330]:
```suggestion
	options := NewTimeoutOptions(ctx, opts...)
```

File: docker.go [hunk 15]
```diff
@@ -343,6 +364,26 @@ func (c *DockerContainer) Terminate(ctx context.Context) error {
...
+
+	defer client.Close()
+
+	// Best effort to remove all volumes.
+	for _, volume := range defaultOptions.Volumes {
+		if errRemove := client.VolumeRemove(ctx, volume, true); errRemove != nil {
+			errs = append(errs, fmt.Errorf("volume remove %q: %w", volume, errRemove))
+		}
+	}
+
```
Comment (stevenh) [inline docker.go:386]: suggest: leverage the Cleanup method.
```suggestion
	if err := options.Cleanup(ctx); err != nil {
		errs = append(errs, err)
	}
```

File: cleanup.go [hunk 16]
```diff
@@ -2,35 +2,37 @@ package testcontainers
...
-	timeout *time.Duration
-	volumes []string
+// DefaultTimeout for termination
+var DefaultTimeout = 10 * time.Second
+
+// TerminateOptions is a type that holds the options for terminating a container.
+type TerminateOptions struct {
+	Context context.Context
+	Timeout *time.Duration
+	Volumes []string
```
Comment (stevenh) [inline cleanup.go:17]: suggestion: use a constructor and methods to ensure consistent behaviour with minimum effort from different types.

```suggestion
// TerminateOptions is a type that holds the options for terminating a container.
type TerminateOptions struct {
        stopTimeout *time.Duration
        volumes []string
        ctx context.Context
}

// NewTerminateOptions returns a fully initialised TerminateOptions.
// Defaults: StopTimeout: 10 seconds.
func NewTerminateOptions(ctx context.Context, opts ...TerminateOption) *TerminateOptions {
        stopTimeout := 10 * time.Second
	options := &TerminateOptions{
		stopTimeout: &stopTimeout,
		ctx: ctx,
	}

	for _, opt := range opts {
		opt(options)
	}

	return options
}

// Context returns the context to use duration a Terminate.
func (o *TerminateOptions) Context() context.Context {
        return o.ctx
}

// StopTimeout returns the stop timeout to use duration a Terminate.
func (o *TerminateOptions) StopTimeout() time.Duration {
        return o.stopTimeout
}

// Cleanup performs any clean up needed.
func (o *TerminateOptions) Cleanup() error {
	// TODO: simplify this when when perform the client refactor.
	if len(o.volumes) == 0 {
		return nil
	}

	client, err := NewDockerClientWithOpts(o.ctx)
	if err != nil {
		return fmt.Errorf("docker client: %w", err)
	}
	defer client.Close()

	// Best effort to remove all volumes.
	var err []errors
	for _, volume := range o.volumes {
		if errRemove := client.VolumeRemove(o.ctx, volume, true); errRemove != nil {
			errs = append(errs, fmt.Errorf("volume remove %q: %w", volume, errRemove))
		}
	}

	return errors.Join(errs...)

**Review by stevenh — CHANGES_REQUESTED** [review]
Sorry to be a pain, but looking at what you've done with this, we should actually make the fields private with methods to avoid any ambiguity as to their use by other types, see suggestion in `docker.go`.

File: cleanup.go [hunk 17]
```diff
@@ -8,16 +8,63 @@ import (
...
+	stopTimeout *time.Duration
+	volumes     []string
 }
 
 // TerminateOption is a type that represents an option for terminating a container.
 type TerminateOption func(*terminateOptions)
 
+// NewTerminateOptions returns a fully initialised TerminateOptions.
+// Defaults: StopTimeout: 10 seconds.
+func NewTerminateOptions(ctx context.Context, opts ...TerminateOption) *terminateOptions {
```
Comment (stevenh) [inline cleanup.go:26]: suggestion: The return type should be exported, as per my previous suggestion. Just keeping fields private ensures the behaviour is consistent.

File: docker.go [hunk 18]
```diff
@@ -343,6 +356,10 @@ func (c *DockerContainer) Terminate(ctx context.Context) error {
 	c.sessionID = ""
 	c.isRunning = false
 
+	if err := options.Cleanup(); err != nil {
```
Comment (stevenh) [inline docker.go:359]: bug: this `err` definition is actually shadowing the existing variable.
```suggestion
	if err = options.Cleanup(); err != nil {
```

File: docker.go [hunk 19]
```diff
@@ -296,19 +296,32 @@ func (c *DockerContainer) Stop(ctx context.Context, timeout *time.Duration) erro
...
+	}
+}
+
+// WithTerminateVolumes is a functional option that sets the volumes for the container termination.
+func WithTerminateVolumes(volumes ...string) TerminateOption {
+	return func(opts *terminateOptions) {
+		opts.volumes = volumes
+	}
+}
+
```
Comment (stevenh) [inline docker.go:312]: bug: These are already covered by `StopTimeout` and `RemoveVolumes` in `cleanup.go`.
```suggestion
```

File: cleanup.go [hunk 20]
```diff
@@ -8,29 +8,76 @@ import (
 	"time"
 )
 
-// terminateOptions is a type that holds the options for terminating a container.
-type terminateOptions struct {
-	ctx     context.Context
-	timeout *time.Duration
-	volumes []string
+// DefaultTimeout for termination
+var DefaultTimeout = 10 * time.Second
```
Comment (stevenh) [inline cleanup.go:12]: bug: sorry mentioned this previously but missed this in the last review, with it being exposed it can introduce race condition.
```suggestion
```

File: cleanup.go [hunk 21]
```diff
@@ -8,29 +8,76 @@ import (
...
 
 // TerminateOption is a type that represents an option for terminating a container.
-type TerminateOption func(*terminateOptions)
+type TerminateOption func(*TerminateOptions)
+
+// NewTerminateOptions returns a fully initialised TerminateOptions.
+// Defaults: StopTimeout: 10 seconds.
+func NewTerminateOptions(ctx context.Context, opts ...TerminateOption) *TerminateOptions {
+	options := &TerminateOptions{
+		stopTimeout: &DefaultTimeout,
```
Comment (stevenh) [inline cleanup.go:28]: suggestion: make local to fix the race condition mentioned above.
```suggestion
	timeout := time.Second * 10
	options := &TerminateOptions{
		stopTimeout: &timeout,
```

**Review by stevenh — CHANGES_REQUESTED** [review]
Sorry missed a bug in the last review

**Review by stevenh — APPROVED** [review]
Thanks for all your work on this, looks good to me.

File: cleanup.go [hunk 22]
```diff
@@ -8,29 +8,74 @@ import (
...
+		stopTimeout: &timeout,
+		ctx:         ctx,
+	}
+	for _, opt := range opts {
+		opt(options)
+	}
+	return options
+}
+
+// Context returns the context to use duration a Terminate.
```
Comment (mdelapenya) [inline cleanup.go:35]: typo: 
```suggestion
// Context returns the context to use during a Terminate.
```

File: cleanup.go [hunk 23]
```diff
@@ -8,29 +8,74 @@ import (
...
+func (o *TerminateOptions) Context() context.Context {
+	return o.ctx
+}
+
+// StopTimeout returns the stop timeout to use duration a Terminate.
```
Comment (mdelapenya) [inline cleanup.go:40]: typo
```suggestion
// StopTimeout returns the stop timeout to use during a Terminate.
```

File: docker_test.go (hunk 11 above)
Comment (mdelapenya) [reply docker_test.go:1122]: I think is not addressed yet. Could you add the mode?

File: cleanup.go [hunk 24]
```diff
@@ -8,29 +8,74 @@ import (
...
+	volumes     []string
 }
 
 // TerminateOption is a type that represents an option for terminating a container.
-type TerminateOption func(*terminateOptions)
+type TerminateOption func(*TerminateOptions)
+
+// NewTerminateOptions returns a fully initialised TerminateOptions.
+// Defaults: StopTimeout: 10 seconds.
+func NewTerminateOptions(ctx context.Context, opts ...TerminateOption) *TerminateOptions {
```
Comment (mdelapenya) [inline cleanup.go:23]: docs: can we make sure we include the new options in the docs site?

**Review by mdelapenya — COMMENTED** [review]
I'm still on PTO, being back Jan 2nd, but had time between sweets and meals to take a quick look.

Just a few comments to be addressed, other than that, will approve once resolved.

Thank you both for your time and effort here

File: cleanup.go (hunk 24 above)
Comment (moogacs) [reply cleanup.go:23]: i've added doc, with consideration it will be released in next RC, lmk if it's not in the right place

**Review by stevenh — APPROVED** [review]

File: docs/features/common_functional_options.md [hunk 25]
```diff
@@ -165,3 +165,41 @@ The above example is updating the predefined command of the image, **appending**
 
 !!!info
     This can't be used to replace the command, only to append options.
+
+#### NewTerminateOptions
```
Comment (mdelapenya) [inline docs/features/common_functional_options.md:169]: I'm sorry but this is not properly documented on our docs 😢 

This file is for the functional options that affect the `Run` function in the modules API.

I'd add this here: https://golang.testcontainers.org/features/garbage_collector/#terminate-function if possible.

Comment (moogacs) [reply docs/features/common_functional_options.md:169]: no worries, i was almost certain i added it in the writing place 😅, will do update it thank you

File: docs/features/garbage_collector.md [hunk 26]
```diff
@@ -17,6 +17,45 @@ The primary method is to use the `Terminate(context.Context)` function that is
 available when a container is created. Use `defer` to ensure that it is called
 on test completion.
 
+#### NewTerminateOptions
+
+- Since testcontainers-go <a href="https://github.com/testcontainers/testcontainers-go/releases/tag/v0.35.0"><span class="tc-version">:material-tag: v0.35.0</span></a>
```
Comment (mdelapenya) [inline docs/features/garbage_collector.md:22]: Again not documented on our side, as the placeholder I'm suggesting will be replaced by the release script with the version we use (it'll be `0.35.0` but imagine we change our minds and decide to release `1.0.0`).

This is our bad, so I'm going to add it to the contributing docs asap 🙏. Please forgive me on this ping-pong.

```suggestion
- Not available until the next release of testcontainers-go <a href="https://github.com/testcontainers/testcontainers-go"><span class="tc-version">:material-tag: main</span></a>
```

File: docs/features/garbage_collector.md [hunk 27]
```diff
@@ -17,6 +17,45 @@ The primary method is to use the `Terminate(context.Context)` function that is
 available when a container is created. Use `defer` to ensure that it is called
 on test completion.
```
Comment (mdelapenya) [inline docs/features/garbage_collector.md:19]:
```suggestion

The `Terminate` function can be customised with termination options to determine how a container is removed: termination timeout, and the ability to remove container volumes are supported at the moment. You can build the default options using the `testcontainers.NewTerminationOptions` function.

```

**Review by mdelapenya — APPROVED** [review]
LGTM, approving it although I added two docs changes.

Will merge this PR right after they are addressed and the CI passes 🤞 

Thank you for your work here 🙇

File: docs/features/garbage_collector.md [hunk 28]
```diff
@@ -17,6 +17,47 @@ The primary method is to use the `Terminate(context.Context)` function that is
...
 
+The `Terminate` function can be customised with termination options to determine how a container is removed: termination timeout, and the ability to remove container volumes are supported at the moment. You can build the default options using the `testcontainers.NewTerminationOptions` function.
+
+#### NewTerminateOptions
+
+- Not available until the next release of testcontainers-go <a href="https://github.com/testcontainers/testcontainers-go"><span class="tc-version">:material-tag: main</span></a>
+
+If you want to attach option to container termination, you can use the `testcontainer.NewTerminateOptions(ctx context.Context, opts ...TerminateOption) *TerminateOptions` option, which receives a TerminateOption as parameter, creating custom termination options to be passed on the container termination.
```
Comment (mdelapenya) [inline docs/features/garbage_collector.md:26]:
```suggestion
If you want to attach option to container termination, you can use the `testcontainers.NewTerminateOptions(ctx context.Context, opts ...TerminateOption) *TerminateOptions` option, which receives a TerminateOption as parameter, creating custom termination options to be passed on the container termination.
```

**Review by mdelapenya — APPROVED** [review]

**Comment by mdelapenya** [comment]
Merged, thank you @moogacs and @stevenh for your time reviewing this PR

File: container.go (hunk 1 above)
Comment (ash2k) [inline container.go:53]: Hm, I wonder why we need a separate parameter for the timeout when a context is already provided to `Terminate()`?

As a user, I expect this to work fine. Does this code not do what is expected?

```go
ctx, _ := context.WithTimeout(context.Background(), time.Second*10)
c.Terminate()
```

If a `Duration` value is needed (vs reacting to a channel close), `context.Background().Deadline()` returns a deadline `Time`. It seems that there is no need for a new API.
